### PR TITLE
issue #658 - split up travis tests, make docker independent

### PR DESCRIPTION
**Pull request by s3bs**
* split up py27 and py35 sims into indepentent tests
* fold travis output for better readability
* add flake8 test (marked with --exit-zero otherwise cocotb would fail)

see: https://travis-ci.com/s3bs/cocotb/builds/83579077

building our own version of iverilog 
the addition of flake8 is meant to start a discussion around pep8. 
#658

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,43 @@
...
-script:
-  - docker build .
-
+jobs:
+  include:
+    - stage: PythonTests
+      install: pip install flake8
+      script: flake8 --exit-zero cocotb/
+    - stage: SimTests
+      # env only here for easier visibiliyt in travis-ci
```
Comment (eric-wieser) [inline .travis.yml]: typo

File: tests/Makefile [hunk 2]
```diff
@@ -34,7 +34,15 @@ REGRESSIONS :=  $(shell ls test_cases/)
 all: $(REGRESSIONS)
 
 $(REGRESSIONS):
+ifeq ($(TRAVIS),true)
+	@echo "travs_fold:start:$@"
+endif
 	cd test_cases/$@ && $(MAKE)
+ifeq ($(TRAVIS),true)
+	@echo "travs_fold:end:$@"
```
Comment (eric-wieser) [inline tests/Makefile]: Is there an `i` mising here?

Comment (s3bs) [reply tests/Makefile]: thanks. I don't even know how that ever worked then.

File: .travis.yml (hunk 1 above)
Comment (s3bs) [reply .travis.yml]: thanks

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,41 @@
...
+    rm -rf /var/lib/apt/lists/* && \
+    apt-get clean && \
+    pip install --upgrade pip && \
+    g++ --version 
+
+# Icarus Verilog 
+ENV ICARUS_VERILOG_VERSION=${ICARUS_VERILOG_VERSION} 
+WORKDIR /usr/src/iverilog
+RUN git clone https://github.com/steveicarus/iverilog.git . && \
+    git checkout v${ICARUS_VERILOG_VERSION} && \
```
Comment (imphil) [inline Dockerfile]: You might want to make this
`git clone https://github.com/steveicarus/iverilog.git --depth=1 -b v${ICARUS_VERILOG_VERSION}`
for faster clones.

File: .travis.yml [hunk 4]
```diff
@@ -1,10 +1,43 @@
...
+    - stage: SimTests
+      # env only here for easier visibility in travis-ci
+      env:
+        - COCOTB_PY35
+      before_install: *docker_prep 
+      script:
+        - echo 'Running cocotb tests with python 35' ... && echo -en 'travis_fold:start:cocotb_py35'
+        - docker exec -it -e TRAVIS=true -e COCOTB=/build-py3/src cocotb bash -lc 'set -x; source /build-py3/venv/bin/activate; make -C /build-py3/src test'
+        - echo 'travis_fold:end:cocotb_py35'
+
```
Comment (imphil) [inline .travis.yml]: Please add a newline at the end of the file.

File: .travis.yml [hunk 5]
```diff
@@ -1,10 +1,43 @@
...
+        - echo 'Running cocotb tests with python 27 ...' && echo 'travis_fold:start:cocotb_py27' 
+        - docker exec -it -e TRAVIS=true -e COCOTB=/build-py2/src cocotb bash -lc 'source /build-py2/venv/bin/activate; make -C /build-py2/src test'
+        - echo -e 'travis_fold:end:cocotb_py27'
+    - stage: SimTests
+      # env only here for easier visibility in travis-ci
+      env:
+        - COCOTB_PY35
+      before_install: *docker_prep 
+      script:
+        - echo 'Running cocotb tests with python 35' ... && echo -en 'travis_fold:start:cocotb_py35'
```
Comment (imphil) [inline .travis.yml]: nit: ... "with Python 3.5"

You also want to include the "..." into the single quotes (or skip them alltogether), and make the second echo a new command in the travis ci file.

File: .travis.yml [hunk 6]
```diff
@@ -1,10 +1,43 @@
...
+      # env only here for easier visibility in travis-ci
+      env:
+        - COCOTB_PY27
+      before_install: &docker_prep
+        - docker build -t cocotb .
+        - docker images -a
+        - docker run -d --name cocotb cocotb tail -f /dev/null
+        - docker ps -a
+      script:
+        - echo 'Running cocotb tests with python 27 ...' && echo 'travis_fold:start:cocotb_py27'
```
Comment (imphil) [inline .travis.yml]: nit: ... "with Python 2.7"

File: .travis.yml [hunk 7]
```diff
@@ -1,10 +1,43 @@
 # See Dockerfile for the full build instructions
 sudo: required
-language: cpp
+language: python
+cache: pip
+
+#mainly for flake8 as everything else runs in docker
```
Comment (imphil) [inline .travis.yml]: nit: "# mainly"

File: .travis.yml [hunk 8]
```diff
@@ -1,10 +1,43 @@
...
+python: 
+  - 3.5
+
+git:
+  quiet: true
```
Comment (imphil) [inline .travis.yml]: I'm not a huge fan of disabling any output in CI, it makes it much harder to diagnose CI failures. But feel free to override me here :-)

File: .travis.yml [hunk 9]
```diff
@@ -1,10 +1,43 @@
...
 
 services:
   - docker
 
-script:
-  - docker build .
-
+jobs:
+  include:
+    - stage: PythonTests
```
Comment (imphil) [inline .travis.yml]: Also the "SimTests" are using Python, making the name "PythonTests" a bit ambiguous in my eyes. How about "lint" or "pythonlint"?

File: Dockerfile [hunk 10]
```diff
@@ -1,19 +1,41 @@
...
-RUN bash -lc 'virtualenv /build-py2/venv; source /build-py2/venv/bin/activate; pip install coverage xunitparser; make -C /build-py2/src test'
+RUN apt-get -qq update && apt-get -qq install -y --no-install-recommends \
+    wget git gperf make autoconf g++ flex bison \
+    python2.7-dev python3-dev\
+    python-pip \
+    python3 virtualenv python3-venv && \
+    rm -rf /var/lib/apt/lists/* && \
+    apt-get clean && \
+    pip install --upgrade pip && \
+    g++ --version
```
Comment (imphil) [inline Dockerfile]: In my experience it's more common to have the `&&` in the next line, as done in the Docker documentation (e.g. here: https://docs.docker.com/develop/develop-images/dockerfile_best-practices). Furthermore, having one line per package makes it easier to extend the list in the future by creating clearer diffs (and reducing merge conflicts)

```
RUN apt-get -qq update \
    && apt-get -qq install -y --no-install-recommends \
        wget \
        git \
        gperf \
        make \
        autoconf \
        g++ \
        flex \
        bison \
        python2.7-dev \
        python3-dev\
        python-pip \
        python3 \
        virtualenv \
        python3-venv \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get clean \
    && pip install --upgrade pip \
    && g++ --version 
```

**Comment by themperek** [comment]
@s3bs Will, you be able to address the comments at some point? I would like to add #703 on top.

File: .travis.yml (hunk 5 above)
Comment (s3bs) [reply .travis.yml]: the second echo has to be on the same line otherwise the script will just exit with 0. 
it kinda serves as anchor for the folding label.

**Comment by imphil** [comment]
@s3bs Looks great, thanks! Would you mind folding the commits into one with a nice commit message (see https://github.com/potentialventures/cocotb/blob/master/CONTRIBUTING.md#patch-requirements)? After that's done this can be merged.

**Comment by imphil** [comment]
Nevermind, I've done the squash through the GitHub UI. Thanks @s3bs!
